### DISCUSSION TITLE
MVC 패턴에서 이벤트 바인딩은 어디에서 담당해줘야할까요?

### DISCUSSION THREAD

**Discussion by moonheekim0118** [woowacourse-study/2022-code-review-study#8] [Lotto] [answered]
MVC를 쓴 크루들 다 다르게 사용하고 있는 것 같아요. 일단 1. Controller에서 Model과 View를 조작(이벤트 핸들링)을 하니까, Controller에 이벤트 바인딩/처리를 붙여놓는 경우와,
2. View에 이벤트 바인딩/처리를 붙여놓는 경우가 있네요. 어느쪽이 더 좋다고 생각하세요 ? 저 같은 경우는 View에다가 바인딩을 해놓긴 했거든요. View 가 해당 Element 를 담당하니까요. 이벤트도 처리해줘야한다고 생각했어요. 그런데 그렇게 되면 어쩔 수 없이 Controller의 이벤트 핸들러를  View에 주입해줘야하는 상황이 생기거든요. 흠...어렵네요..

**Comment by lokba** [comment, marked as answer]
저 또한 1번과 2번 방식 모두를 적용해보았는데요. 
제 의견보다는 아래 사진으로 대체해보겠습니다.


<img width="1068" alt="스크린샷 2022-03-03 오후 10 34 45" src="https://user-images.githubusercontent.com/19251499/156592530-a5e32820-1cbd-4d85-bb3c-ac163fef8ab3.png">

**Reply by wonsss** [reply]
> "(view에서 이벤트를 등록한 경우) 이벤트 발생시 컨트롤러에서의 동작은 별도의 함수를 마련, 호출함으로써 해결하면 되구요"

view에서 이벤트를 등록한 경우, 이벤트 발생시 view가 controller의 동작을 실행하게 하려면, 단순하게 생각했을 때 controller의 메서드를 view 내부에서 호출할 것이고 이는 view가 controller에게 의존하게 될 것 같아요. 그러면 서로 너무 의존하게 되는 단점이 발생하네요.

그러면 view가 controller를 직접적으로 호출하지 않으면서, view에서 등록된 이벤트가 감지된 경우 controller에게 어떻게 알릴 수 있을까요? 그래서 로이정님도 `컨트롤러에서의 동작은 별도의 작은 함수를 마련`하라고 알려주신 것 같은데, 그 방법으로는 구체적으로 어떤 것들이 있을까요?

**Reply by lokba** [reply]
네 맞습니다. 
로이님이 controller와 view의 의존성을 끊는 방법에 대해서도 저에게 직접 코멘트를 달아주셨는데요.
customEvent 활용, observer패턴 도입 등이 있었습니다.
다만, 저는 아직 학습을 하지 못한 상태여서 구체적으로는 설명을 못드리겠네요.

**Reply by wonsss** [reply]
> 그러면 view가 controller를 직접적으로 호출하지 않으면서, view에서 등록된 이벤트가 감지된 경우 controller에게 어떻게 알릴 수 있을까요? 그래서 로이정님도 컨트롤러에서의 동작은 별도의 작은 함수를 마련하라고 알려주신 것 같은데, 그 방법으로는 구체적으로 어떤 것들이 있을까요?

[CustomEvent](https://developer.mozilla.org/ko/docs/Web/API/CustomEvent/CustomEvent)와 [disptachEvent](https://developer.mozilla.org/ko/docs/Web/API/EventTarget/dispatchEvent)를 활용하는 방법도 있을 것 같습니다.
view에서는 선택한 Element에 이벤트를 등록하고, 이벤트가 감지되면  CustomEvent 생성자로 `별도의 이벤트이름`과 값을 담은 새 Event 객체를 만들어 다시  해당 Element에 dipatchEvent를 통해 실행시켜 줄 수 있습니다. 이렇게 disptachEvent까지 하여 커스텀이벤트를 실행해주면, Controller에서는 addEventListener하는 이벤트의 대상이 'click'같은 원래 내장된 이벤트가 아니라 아까 만든 커스텀 이벤트의 `별도의 이벤트 이름` 으로 이벤트를 등록할 수 있습니다. 그리고 Controller에서 핸들러가 그 커스텀 이벤트에 대해서도 반응할 수 있습니다. 

이 방법을 사용하면, view가 이벤트를 등록받으면서도 이벤트 발생시 controller를 호출(의존)하지 않게 되는 것 같습니다.

이 방법은 어떤지 의견이 궁금합니다. 그리고 이 외에 다른 방법도 있을 것 같은데 아시는 분 있을까요?

**Reply by moonheekim0118** [reply]
오 ~ CustomEvent를 이렇게 사용하는군요~
저는 개인적으로 Observer패턴을 선호하는 편입니다만..이렇게하면 MVC패턴이 아니라 Flux에 더 가까워지는 것 같아요.
 하나의 상태 Store를 정해놓고 상태가 변경될 경우 View에 Notify 해주면 View가 리렌더링 되는 방식이요.
그러면 `Controller에서 Store 에 상태를 변경 -> Store에서 구독중인 View들에게 Notify -> View 리렌더링` 이렇게 진행되어서 Controller와 View의 의존성을 끊게 되긴 하거든요
